Account for the Darcy Velocity and/or the Fluid Velocity when Determining Outflow Boundaries

danieldouglas92 opened the pull request:
Currently, the `replace_outflow_boundary_ids` function only uses the solid velocity to determine if a boundary should be flagged as an outflow boundary. This PR updates the `replace_outflow_boundary_ids` function to account for the darcy velocity when using the `darcy field` compositional field advection method. 

Based on a conversation with @jdannberg this will also need to be extended to account for the fluid velocity when simulating melt transport, but I figured I would open the PR with what I have first to make sure that the implementation looks good and then add the other case afterwards.

This PR addresses the last point in Issue #4748.

danieldouglas92 committed:
account for the darcy and fluid velocities when computing outflow boundaries

tjhei commented on source/simulator/helper_functions.cc:
```diff
@@ -2275,13 +2275,28 @@ namespace aspect
+    MaterialModel::MaterialModelInputs<dim> in(fe_face_values.n_quadrature_points, introspection.n_compositional_fields);
+    MaterialModel::MaterialModelOutputs<dim> out(fe_face_values.n_quadrature_points, introspection.n_compositional_fields);
+    MeltHandler<dim>::create_material_model_outputs(out);
+    MaterialModel::MeltOutputs<dim> *fluid_out = nullptr;
```
I don't see where this is being initialized but you are reading from this below.

danieldouglas92 commented:
Thanks for the quick review @tjhei! I've added another if statement which accounts for the fluid velocity used by the melt transport model when determining whether a boundary is an outflow boundary

jdannberg commented on source/simulator/helper_functions.cc:
```diff
@@ -2275,13 +2275,28 @@ namespace aspect
     std::vector<Tensor<1,dim>> face_current_velocity_values (fe_face_values.n_quadrature_points);
+    std::vector<Tensor<1,dim>> face_current_fluid_velocity_values (fe_face_values.n_quadrature_points);
+
+    MaterialModel::MaterialModelInputs<dim> in(fe_face_values.n_quadrature_points, introspection.n_compositional_fields);
+    MaterialModel::MaterialModelOutputs<dim> out(fe_face_values.n_quadrature_points, introspection.n_compositional_fields);
+    MeltHandler<dim>::create_material_model_outputs(out);
+    MaterialModel::MeltOutputs<dim> *fluid_out = out.template get_additional_output<MaterialModel::MeltOutputs<dim>>();;
```
You are using the material model outputs below, but you do not evaluate the material model. So I don't think the values will be right.

jdannberg replied:
To call evaluate, you first need to call `in.reinit(...)`, and you need to do that within the loop over all faces (below line 2326). Each material model inputs object has only as many points as there are on one face of one mesh cell, so each time you go to a new face, you need to update the values. Then once you have the right values in the material model inputs, you can call evaluate to get the right values in the outputs. Then you can use them.

jdannberg reviewed (commented):
Thank you for taking a stab at this!
I think this goes in the right direction. But I think it would be good to have a test case that shows that this actually works (one with darcy flow where the solid flows in and the melt flow out, for example).

danieldouglas92 commented on source/simulator/helper_functions.cc:
```diff
@@ -2325,12 +2344,35 @@ namespace aspect
 
                 // ... and change the boundary id of any outflow boundary faces.
-                if (integrated_flow > 0)
+                if (integrated_solid_flow > 0 || integrated_darcy_flow > 0 || integrated_fluid_flow > 0)
```
This makes sense! I've changed this so that they all need to be outflowing to update the cell.

tjhei commented on source/simulator/helper_functions.cc:
```diff
@@ -2311,10 +2340,22 @@ namespace aspect
+                    material_model->evaluate(in, out);
+                  }
+
+                if (parameters.include_melt_transport)
+                  integrated_fluid_flow = 0;
```
these two lines seem redundant.

danieldouglas92 replied:
I do this because by default (no melt transport) `integrated_fluid_flow = std::numeric_limits<double>::max()` so that the if statement at line 2400 still works in solid only models. If there is melt though, then I need to set `integrated_fluid_flow = 0` otherwise line 2388:
```
integrated_fluid_flow += (face_current_fluid_velocity_values[q] * fe_face_values.normal_vector(q)) *
                                          fe_face_values.JxW(q);
```
will always be infinity. This was the only way I could think to get around this issue, but if you have a better approach that would be great!

danieldouglas92 replied:
The same is done for when a field is advected using the darcy advection method at line 2352

tjhei replied:
Oh, I see. I would put that into the initialization directly then.

danieldouglas92 replied:
Ah yes! That will make this look much cleaner

danieldouglas92 commented:
@jdannberg @tjhei John and I just talked about what this PR is now doing given that the condition for flagging a cell as an outflow cell requires outflow of all three velocities and we're wondering if we should just scrap this PR and jump right into making it composition dependent.

The issue that this PR was trying to address was that in a model with two-phases the solid phase could be inflowing while the fluid phase is outflowing, this is why I had initially implemented line 2388 as an OR condition and not an AND condition. Since it is an AND condition now (to not violate the boundary condition for the solid as it inflows), the original issue that this PR is trying to address is no longer solved.

Additionally, in the rare case where the solid is outflowing and the fluid is inflowing, the cell will not be flagged as outflow because both phases are not outflowing, which is an issue that wasn't present prior to this PR.

Am I missing something that makes what I'm currently trying to do still valuable? If not, I will close this PR and work on a composition dependent implementation.

jdannberg commented:
I think it makes sense to directly go to implementing this in dependence of the compositional field type. But I think you can just work from what you have here. The only changes that would be required are

(1) add an input parameter to the replace_outflow_boundary_ids function that tells the function if it computes boundary ids for the temperature or compositional field, and which field, 
(2) to modify what you have right now so that it checks for the compositional field type (or temperature), and if it's the porosity (or a `melt field`) in a model with melt transport, or a `darcy field`, you pick the appropriate velocity to determine if model has outflow (rather than the current AND/OR statement) 
(3) change core.cc so that replace_outflow_boundary_ids and restore_outflow_boundary_ids is called within the loop over all compositional fields (Do [this](https://github.com/geodynamics/aspect/blob/dcf512bf81fb45e7cb0b1db8849db1a89583ea28/source/simulator/core.cc#L736-L737) and [this](https://github.com/geodynamics/aspect/blob/dcf512bf81fb45e7cb0b1db8849db1a89583ea28/source/simulator/core.cc#L765-L766) in a loop for all compositional field and also for the temperature)

danieldouglas92 commented:
@jdannberg Here's my initial go at making this composition dependent. I wasn't sure if I needed to make `restore_outflow_boundary_ids` composition dependent so I left this function as it was for now to see how the testers did.

jdannberg commented on source/simulator/core.cc:
```diff
@@ -695,7 +695,7 @@ namespace aspect
     const unsigned int boundary_id_offset = 128;
     if (!boundary_temperature_manager.allows_fixed_temperature_on_outflow_boundaries())
-      replace_outflow_boundary_ids(boundary_id_offset);
+      replace_outflow_boundary_ids(boundary_id_offset, false, 0);
```
You can even set the index to an invalid unsigned int here (since it is not going to be used in case of temperature).

jdannberg commented on source/simulator/helper_functions.cc:
```diff
@@ -2271,17 +2271,53 @@ namespace aspect
+            consider_darcy_velocity = true;
+            porosity_idx = introspection.find_composition_type(CompositionalFieldDescription::porosity);
+          }
+        if (introspection.compositional_field_methods[composition_index] == Parameters<dim>::AdvectionFieldMethod::fem_melt_field)
+          consider_melt_velocity = true;
```
The porosity itself is not an `fem_melt_field` (see for example test advect_field_with_melt_velocity.prm, which is the only one I have found that actually uses the melt fields), so `consider_melt_velocity` should additionally be true if both the model includes melt transport and the composition with index `composition_index` is a field of type porosity.

jdannberg commented on source/simulator/helper_functions.cc:
```diff
@@ -2325,8 +2367,34 @@ namespace aspect
+
+                    else if (consider_melt_velocity)
+                      {
+                        const FEValuesExtractors::Vector ex_u_f = introspection.variable("fluid velocity").extractor_vector();
+                        fe_face_values[ex_u_f].get_function_values(current_linearization_point, face_current_fluid_velocity_values);
```
You can move this part out of the loop over all q, up to where the same is done for the solid velocity. This way, you only do this once per face, rather than once per quadrature point on the face.

jdannberg commented on source/simulator/helper_functions.cc:
```diff
@@ -2311,11 +2347,17 @@ namespace aspect
-
                 // ... check if the face is an outflow boundary by integrating the normal velocities
                 // (flux through the boundary) as: int u*n ds = Sum_q u(x_q)*n(x_q) JxW(x_q)...
+                // do this for the solid velocity, the darcy velocity, and/or the fluid velocity.
```
```suggestion
                // do this for the solid velocity, the darcy velocity, or the fluid velocity.
```

jdannberg reviewed (commented):
I think this should work!

There's one case (porosity in the case of melt transport) which I think does not work correctly right now (see other comments), but I think in general this should now apply the different velocities correctly at the boundary. You also do not need to change `restore_outflow_boundary_ids` because that just resets the values to what they were originally. 

Did you already check the tests to see if the results look right? Because it would make sense that some tests are different now (since you changed the functionality).

danieldouglas92 commented:
Thanks for the review @jdannberg! The tests are failing because the boundary_ids are exceeding 128, working on fixing this bug now but hopefully after that is fixed this is good to go!

danieldouglas92 committed:
Add changelog

danieldouglas92 commented:
Working

danieldouglas92 committed:
add test case

danieldouglas92 commented:
While trying to come up with a test I ran into an issue with the case where `consider_darcy_velocity = true`. When this part of the code:

```
                if (consider_darcy_velocity)
                  {
                    in.reinit(fe_face_values, cell, introspection, solution);
                    material_model->evaluate(in, out);
                    fluid_out = out.template get_additional_output<MaterialModel::MeltOutputs<dim>>();
                  }
```

gets called (specifically `material_model->evaluate(in, out);`), the model terminates because the temperature, pressure etc. are all 0.

A similar error happens for the same test if porosity is not advected with the `darcy field` advection method, but instead with the fluid velocity calculated by the melt model. In core.cc at when [this line](https://github.com/geodynamics/aspect/blob/main/source/simulator/core.cc#L768) gets called, this triggers the same error when the code `this->get_material_model().evaluate(material_model_inputs, material_model_outputs);` in [melt.cc](https://github.com/geodynamics/aspect/blob/main/source/simulator/melt.cc#L1523) gets called and terminates due to the temperature, pressure etc. all being 0.

I'm not sure what to make of this, as for the fluid velocity case this all happening in functions that I didn't modify in this PR. The only connection between what I changed and the fluid velocity case is through the variable `new_current_constraints`, but it's not clear to me how what I've done would propagate through into this variable. I've added a commit just now which includes the test that I was working on in an unfinished state.

jdannberg commented:
I think I don't yet understand what exactly the problem is you're describing. 
What exactly is the error that makes the model crash? In other words, why is it a problem if temperature, pressure etc. are all zero? (I get that it's not what you put in, but that's a different issue.) At which point does it crash? (for example, is the temperature zero because it's before the initial conditions have been set, or should they have been set and for some reason are zero now?)

The line 
`in.reinit(fe_face_values, cell, introspection, solution)`
 just takes whatever values are in `solution` and puts those into `in`, so what do you expect to be in the solution vector at that point in time where it crashes and is that the actually the case?

Does your input file crash if you run it on ASPECT main? Or is the crash actually related to the changes you made?

danieldouglas92 commented:
@jdannberg Yes sorry for not including more background detail on this. My first instinct was, as you pointed out could be an issue, was that I was trying to use the material model inputs before they were initialized and so were set to 0. 

For the Darcy velocity case, this error does only happens on this branch, and does not happen on main. For the fluid velocity case it happens on both main and on my branch. It isn't surprising that this doesn't happen for the Darcy velocity case on main though, because the line `material_model->evaluate(in, out);` isn't there.

I think I'm just now figuring out why this is happening. The test `no_dirichlet_on_outflow_melt`, and most of the tests involving melt, use the melting rate material model from`rising_melt_blob.cc` which explicitly defines what the outputs should be regardless of the modeled temperature and pressure. This test that I've created relies on using the visco plastic material model to determine the material outputs based on the pressure and temperature, but these are not initialized yet when `replace_outflow_boundary_ids` is called.

Of course I could probably circumnavigate this by using `rising_melt_blob.cc` (I'll double check that this actually works soon) in this test, but this wouldn't actually fix the issue. If you agree that my view on the problem is right, what do you think would be the best way to get around this? I have some ideas for using a reference pressure or temperature only when the function is called per-initialization of the material model inputs, but they seem kind of hacky.

danieldouglas92 commented:
@jdannberg I got around the issue by properly setting the initial adiabatic conditions! Now the tests are not crashing, but they are producing some behaviour that is really bizarre. For both cases, the test simulates a blob of porosity in the middle of the model which flows up to outflow through the top boundary. At the top boundary, there is a solid field called `solid_phase` which is being advected downward at 10 cm/yr due to the prescribed velocity boundary conditions. The hope is that this will allow the `solid_phase` composition to be correctly handled as an inflow while the `porosity` composition will be allowed to outflow at the same time. This is what the initial conditions look like

<img width="678" alt="image" src="https://github.com/user-attachments/assets/10873ba6-6239-452d-b47e-8e97d685bc41">

For the fluid velocity case, when the blob hits the top boundary the porosity starts to increase across the whole domain and the model crashes. This is what happens to the `porosity` composition:
<img width="765" alt="image" src="https://github.com/user-attachments/assets/659f3e85-9b02-4cce-bcbf-e9bee6dc08ce">

For the Darcy velocity case though, this seems to be ok.
<img width="771" alt="image" src="https://github.com/user-attachments/assets/7a76d2e7-ee16-4c8f-ba15-d4c7100d8aad">

I'm not exactly sure what is causing the fluid velocity case to crash. The model hangs when run in debug mode, and terminates with an MPI error in release mode.

jdannberg commented:
For the first problem with the pressure and temperature: Are you saying you need to set the adiabatic conditions to a function, and if you use initial profile, it does not work? If that’s the case, we have many places in the material models where we check if `this->get_adiabatic_conditions().is_initialized()` and then do something different if it is not initialized. So whichever computation of material properties can’t deal with adiabatic pressure and temperature being unavailable, that’s what would have to be changed. But that does not directly relate to your PR, it’s just something we should fix independently. Since you say it’s broken on main, could you make an issue describing the problem in detail?

For the blob: That does seem strange. And I would say, even for the Darcy velocity, it does not seem right. It looks like there is a small blue line at the top of the box, so the outflow boundary conditions are not actually recognized correctly and porosity is still set to zero at the boundary (whereas at the bottom boundary, porosity along the boundary is the same as what it is in the interior). 

For the fluid velocity case: where does it hang in debug mode? Can you run it in a debugger and figure out the line? And it looks like you’ve set up the model to not have melting or freezing, so it really just advects the porosity field? Is there anything else that is weird about the model (pressure, temperature, etc.)? Does it also look like that and crash on main or is it related to the boundary conditions?

And at least the boundary condition part might work for the porosity? It is set to zero along parts of the upper and the whole lower boundary, but not where the blob reached the boundary.

danieldouglas92 commented:
@jdannberg Thanks for following up on this! 

For the first problem, yes even on the main branch if I do not define an `Adiabatic conditions model` (which then uses the `compute profile` model as the default), the model immediately crashes because the temperature and pressure are both 0. When I set the `Adiabatic conditions model` with a function which explicitly defines the pressure and temperature, then the model proceeds.

For the darcy flow case, you are right I didn't notice that there is a thin blue line along the top boundary where the `porosity` is trying to outflow. 

For the fluid velocity case, I wasn't able to found out an exact line unfortunately, but I was
able to narrow it down to this when this [function](https://github.com/geodynamics/aspect/blob/main/source/simulator/solver.cc#L598) is called [here](https://github.com/geodynamics/aspect/blob/main/source/simulator/solver_schemes.cc#L173). When I run it with the debugger it makes it to the end of this function when advecting the `porosity` composition, and then it throws a Trilinos exception. This is the exact message:

```
---------------------------------------------------------
TimerOutput objects finalize timed values printed to the
screen by communicating over MPI in their destructors.
Since an exception is currently uncaught, this
synchronization (and subsequent output) will be skipped
to avoid a possible deadlock.
---------------------------------------------------------
2

----------------------------------------------------
Exception 'ExcTrilinosError(ierr)' on rank 0 on processing: 

--------------------------------------------------------
An error occurred in line <1678> of file </Users/danieldouglas/software/dealii-9.5.2-install/tmp/unpack/deal.II-v9.5.2/source/lac/trilinos_sparse_matrix.cc> in function
    void dealii::TrilinosWrappers::SparseMatrix::add(const size_type, const size_type, const size_type *, const TrilinosScalar *, const bool, const bool)
The violated condition was: 


For historical reasons, many Trilinos functions express errors by
returning specific integer values to indicate certain errors.
Unfortunately, different Trilinos functions often use the same integer
values for different kinds of errors, and in most cases it is also not
documented what each error code actually means. As a consequence, it
is often difficult to say what a particular error (in this case, the
error with integer code '2') represents and how one should fix a code
to avoid it. The best one can often do is to look up the call stack to
see which deal.II function generated the error, and which Trilinos
function the error code had originated from; then look up the Trilinos
source code of that function (for example on github) to see what code
path set that error code. Short of going through all of that, the only
other option is to guess the cause of the error from the context in
which the error appeared.
```

Unfortunately I haven't been able to make much of this error message... I found the error message in the dealii source code but as the error message points out I didn't find it very helpful. If you've encountered this before or have any ideas on how to proceed I would really appreciate it.


Also, here is the stacktrace in case that might be helpful:

```
Stacktrace:
-----------
#0  2   libdeal_II.g.9.5.2.dylib            0x0000000113efae00 _ZN6dealii16TrilinosWrappers12SparseMatrix3addEjjPKjPKdbb + 532: 2   libdeal_II.g.9.5.2.dylib            0x0000000113efae00 _ZN6dealii16TrilinosWrappers12SparseMatrix3addEjjPKjPKdbb 
#1  3   libdeal_II.g.9.5.2.dylib            0x0000000113bde68c _ZNK6dealii17AffineConstraintsIdE26distribute_local_to_globalINS_16TrilinosWrappers17BlockSparseMatrixENS3_3MPI11BlockVectorEEEvRKNS_10FullMatrixIdEERKNS_6VectorIdEERKNSt3__16vectorIjNSF_9allocatorIjEEEERT_RT0_bNSF_17integral_constantIbLb1EEE + 1156: 3   libdeal_II.g.9.5.2.dylib            0x0000000113bde68c _ZNK6dealii17AffineConstraintsIdE26distribute_local_to_globalINS_16TrilinosWrappers17BlockSparseMatrixENS3_3MPI11BlockVectorEEEvRKNS_10FullMatrixIdEERKNS_6VectorIdEERKNSt3__16vectorIjNSF_9allocatorIjEEEERT_RT0_bNSF_17integral_constantIbLb1EEE 
#2  4   aspect-debug                        0x00000001004ecdd4 _ZN6aspect9SimulatorILi2EE37copy_local_to_global_advection_systemERKNS1_14AdvectionFieldERKNS_8internal8Assembly8CopyData15AdvectionSystemILi2EEE + 92: 4   aspect-debug                        0x00000001004ecdd4 _ZN6aspect9SimulatorILi2EE37copy_local_to_global_advection_systemERKNS1_14AdvectionFieldERKNS_8internal8Assembly8CopyData15AdvectionSystemILi2EEE 
#3  5   aspect-debug                        0x0000000100500c2c _ZN6dealii10WorkStream8internal10sequential3runIZN6aspect9SimulatorILi2EE25assemble_advection_systemERKNS6_14AdvectionFieldEEUlRKNS_18TriaActiveIteratorINS_15DoFCellAccessorILi2ELi2ELb0EEEEERNS4_8internal8Assembly7Scratch15AdvectionSystemILi2EEERNSH_8CopyData15AdvectionSystemILi2EEEE_ZNS6_25assemble_advection_systemES9_EUlRKSO_E_NS_16FilteredIteratorISD_EESK_SO_EEvRKT1_RKNS_9std_cxx2013type_identityISW_E4typeET_T0_RKT2_RKT3_ + 212: 5   aspect-debug                        0x0000000100500c2c _ZN6dealii10WorkStream8internal10sequential3runIZN6aspect9SimulatorILi2EE25assemble_advection_systemERKNS6_14AdvectionFieldEEUlRKNS_18TriaActiveIteratorINS_15DoFCellAccessorILi2ELi2ELb0EEEEERNS4_8internal8Assembly7Scratch15AdvectionSystemILi2EEERNSH_8CopyData15AdvectionSystemILi2EEEE_ZNS6_25assemble_advection_systemES9_EUlRKSO_E_NS_16FilteredIteratorISD_EESK_SO_EEvRKT1_RKNS_9std_cxx2013type_identityISW_E4typeET_T0_RKT2_RKT3_ 
#4  6   aspect-debug                        0x00000001004edc24 _ZN6dealii10WorkStream3runIZN6aspect9SimulatorILi2EE25assemble_advection_systemERKNS4_14AdvectionFieldEEUlRKNS_18TriaActiveIteratorINS_15DoFCellAccessorILi2ELi2ELb0EEEEERNS2_8internal8Assembly7Scratch15AdvectionSystemILi2EEERNSF_8CopyData15AdvectionSystemILi2EEEE_ZNS4_25assemble_advection_systemES7_EUlRKSM_E_NS_16FilteredIteratorISB_EESI_SM_EEvRKT1_RKNS_9std_cxx2013type_identityISU_E4typeET_T0_RKT2_RKT3_jj + 220: 6   aspect-debug                        0x00000001004edc24 _ZN6dealii10WorkStream3runIZN6aspect9SimulatorILi2EE25assemble_advection_systemERKNS4_14AdvectionFieldEEUlRKNS_18TriaActiveIteratorINS_15DoFCellAccessorILi2ELi2ELb0EEEEERNS2_8internal8Assembly7Scratch15AdvectionSystemILi2EEERNSF_8CopyData15AdvectionSystemILi2EEEE_ZNS4_25assemble_advection_systemES7_EUlRKSM_E_NS_16FilteredIteratorISB_EESI_SM_EEvRKT1_RKNS_9std_cxx2013type_identityISU_E4typeET_T0_RKT2_RKT3_jj 
#5  7   aspect-debug                        0x00000001004ed634 _ZN6aspect9SimulatorILi2EE25assemble_advection_systemERKNS1_14AdvectionFieldE + 1232: 7   aspect-debug                        0x00000001004ed634 _ZN6aspect9SimulatorILi2EE25assemble_advection_systemERKNS1_14AdvectionFieldE 
#6  8   aspect-debug                        0x0000000100103b34 _ZN6aspect9SimulatorILi2EE30assemble_and_solve_compositionERKNSt3__16vectorIdNS2_9allocatorIdEEEEPS6_ + 936: 8   aspect-debug                        0x0000000100103b34 _ZN6aspect9SimulatorILi2EE30assemble_and_solve_compositionERKNSt3__16vectorIdNS2_9allocatorIdEEEEPS6_ 
#7  9   aspect-debug                        0x0000000100105188 _ZN6aspect9SimulatorILi2EE35solve_iterated_advection_and_stokesEv + 336: 9   aspect-debug                        0x0000000100105188 _ZN6aspect9SimulatorILi2EE35solve_iterated_advection_and_stokesEv 
#8  10  aspect-debug                        0x0000000100ae4998 _ZN6aspect9SimulatorILi2EE14solve_timestepEv + 252: 10  aspect-debug                        0x0000000100ae4998 _ZN6aspect9SimulatorILi2EE14solve_timestepEv 
#9  11  aspect-debug                        0x0000000100ae3498 _ZN6aspect9SimulatorILi2EE3runEv + 892: 11  aspect-debug                        0x0000000100ae3498 _ZN6aspect9SimulatorILi2EE3runEv 
#10  12  aspect-debug                        0x0000000100a633e8 _Z13run_simulatorILi2EEvRKNSt3__112basic_stringIcNS0_11char_traitsIcEENS0_9allocatorIcEEEES8_bbbb + 1076: 12  aspect-debug                        0x0000000100a633e8 _Z13run_simulatorILi2EEvRKNSt3__112basic_stringIcNS0_11char_traitsIcEENS0_9allocatorIcEEEES8_bbbb 
#11  13  aspect-debug                        0x0000000100a62a40 main + 2300: 13  aspect-debug                        0x0000000100a62a40 main 
#12  14  dyld                                0x000000018ea0b154 start + 2476: 14  dyld                                0x000000018ea0b154 start 
--------------------------------------------------------
```

jdannberg commented:
Oh, I see. So everything works fine with the adiabatic conditions as long as you set the adiabatic surface temperature. That's good to hear. 

For the fluid velocity case, for me it worked with a high resolution everywhere:
```
subsection Mesh refinement
  set Initial adaptive refinement              = 0
  set Initial global refinement                = 2
  set Time steps between mesh refinement       = 0
end
```
I am still not sure what the problem is, and it would be good to figure that out, but you could try if your current file also crashes on main, and if so, then at least you know it does not have anything to do with your implementation. Since I had simplified a few more things, I am attaching my input file here.
[original.prm.txt](https://github.com/user-attachments/files/17781446/original.prm.txt)

I also still see the porosity increasing in the box at the end even though I don't think any fluid flows in and there are no reactions. That still seems quite wrong, but at the moment I can't see where that comes from.

Looking at the output, for the fluid flow case the boundary conditions seem correct for the porosity (for the solid_phase field it is hard to tell, because the value at the outflow boundary is the same as the initial condition). 

> I wonder if the bottom boundary looking correct might not be surprising since the solid is also outflowing here?

Yes, I think that makes sense and gives you an idea where to search for the bug.

danieldouglas92 commented:
@jdannberg I'm pretty confident I've figured out what is going on with  these tests. First, when I increase the resolution of the fluid velocity test, the behaviour does drastically improve, and the behaviour of the test using the functionality implemented in this PR relative to the `main` branch definitely suggests that it is working as expected.

On the main branch, because the top boundary is strictly an inflow boundary, the `porosity` pools up and spreads out along the top boundary. With the functionality included in this PR, the porosity is allowed to outflow while the solid inflows, and no pooling happens at the top boundary. Two screenshots from the same model time comparing the behaviour on main vs this PR are shown below.

![image](https://github.com/user-attachments/assets/992bfc67-64a3-4ee0-a1f9-568d618ad024)

![image](https://github.com/user-attachments/assets/9924b065-60be-4d59-a952-f018f88600f2)


For the Darcy velocity test, the issue is that the Darcy velocity depends on the `porosity` compositional field. In this test, when I had set that the composition should be fixed on the `top` and `bottom` boundaries to the `initial composition`, the initial composition has the `porosity` field equal to 0 on both of these boundaries. This means that the `darcy velocity` will ALWAYS be essentially whatever the `solid velocity` is, and in this test case since the solid velocity is NEGATIVE on the top boundary, the porosity will NEVER outflow. The only way to allow the porosity to outflow is for there to BE porosity, and it's never able to become non-zero due to the enforced boundary conditions.

If I define the porosity to be uniform everywhere and create an additional field which is advected with the `darcy field` method (following the setup of `tests/darcy_velocity.prm`), the model behaves as expected and the behaviour is more correct relative to the `main` branch. Below are a few of screenshots comparing the behaviour of this test between the changes in this PR and with `main`.


The first pair of screenshots shows how when using the main branch the `free_fluid` composition is zero at the top boundary since the boundary is strictly an inflow boundary, and also how numerical instabilities arise later in the model.
Note the colorbar scale change in the second screenshot, where the composition blows up.

![image](https://github.com/user-attachments/assets/a124ef65-c817-47b6-857f-2ac4bcf6069d)

![image](https://github.com/user-attachments/assets/6f56b0d2-c20a-4836-8ec8-15a1956f10a3)


The second pair of screenshots shows the behaviour on this branch, where these issues are not present. The composition effectively outflows through the top boundary and avoids blowing up later in the model.

![image](https://github.com/user-attachments/assets/4abaa1ef-90ad-4b4c-88f2-6d06da955051)

![image](https://github.com/user-attachments/assets/6724fd38-061b-4f53-8b23-df3e00e3aaab)


While I'm confident that this is why the Darcy velocity test was not working, the solution of defining a homogeneous `porosity`  is obviously not ideal. The framework that we've put together for using the `darcy field` advection method with the `Reactive Fluid Transport Model` requires the `porosity` compositional field to be dynamic and most models likely will not start with non-zero porosity on all boundaries.

jdannberg commented:
Hi Daniel, 

Sorry for taking such a long time to get back to this. 
I think you’re right and the original test case you made for the Darcy flow was just not the right setup for testing this feature. So I agree that it makes sense to just use a better test case. I would suggest something like a linear increase in porosity from left to right so that some parts of the boundary actually will have a prescribed porosity because the solid velocity has a bigger magnitude than the Darcy velocity, and some will have outflow of melt and then the porosity should not be prescribed. 

But I think with Darcy flow, it is possible for porosity to flow into a region where no/very low porosity was before and also across the top boundary. I took your test and reduced the permeability a bit, increased the adaptive refinement by 1, and increased the stabilization (see below). And the porosity does flow out at the top (see attacked inamges). I discussed this with @gassmoeller, and his point was this: We do not solve an equation for the Darcy flow, it is a value we compute point by point. So the Darcy velocity at any given point is independent of the Darcy velocity at the point next to it. This means that the Darcy field can be transported upwards at the point just below the top boundary even if the top boundary has a net velocity for the Darcy field that points downward. This means even though the porosity is prescribed to be almost zero at the top boundary, and the flow for the Darcy field should therefore be inward, the porosity can still flow out. The value at the boundary is then just overwritten by the boundary conditions, which is a bit annoying since it can cause oscillations in the porosity field. But for your application, this should work. 

With a finite-element field where we solve an equation for the porosity, this would not work because the velocity would have to be divergence-free/mass would need to be conserved. And if it’s divergence free, that kind of flow field where the velocity at the top of a cell points downward, but at the bottom it points upward would not ever be a solution of the equation. So that’s something to keep in mind: When you advect material with the Darcy velocity, you have no guarantee that its mass is being conserved. 

And the point about the stabilization: I think a general problem with the Darcy fields is that the equation naturally leads to very big gradients in porosity (since regions with a lot of porosity rise more quickly, the top of every rising blob tends to have a sharp jump in porosity). Therefore, you might need more stabilization to avoid over- and undershoots. At least with your specific test case, this helped a lot. You can set the beta and cR stabilization parameters for each individual field (see their documentation), so for your application models, I would just use these very large values for the Darcy fields, but not any other fields. The faster the Darcy velocity, the more stabilization you’ll need. This might also improve nonlinear solver convergence. For the test case here, I used 

```
subsection Discretization
  subsection Stabilization parameters
    set beta = 0.4
    set cR = 2
    set Use artificial viscosity smoothing = true
  end
end
```

and that helped quite a bit. 

So for this PR, I would say: Change the Darcy flow test case to one that actually illustrates that everything works as intended, and then we can at least get this PR merged and worry about any other problems later. 

![0](https://github.com/user-attachments/assets/8778c894-4f42-49a1-95b3-44f5e71f373d)
![1](https://github.com/user-attachments/assets/d431e089-71e1-48b2-9c2d-bfaf14d4d563)
![2](https://github.com/user-attachments/assets/d6b7aead-c2d9-4b05-a146-16f3533d7c76)
![3](https://github.com/user-attachments/assets/bcdf4db2-bc27-424b-925f-91978d7cd0e2)
![4](https://github.com/user-attachments/assets/a6991c4e-0feb-4479-a78f-33a8f98073c2)
![5](https://github.com/user-attachments/assets/d8a4dc5a-b974-498c-a54b-13eaea1e4ec4)
![6](https://github.com/user-attachments/assets/71902a4c-8628-42d9-bbe7-cac48651a7f9)
![7](https://github.com/user-attachments/assets/4d6a30fb-cb39-496c-bd89-15afa8a44cde)

danieldouglas92 commented:
Hey @jdannberg thanks for the detailed explanation and the confirmation that what is in this PR seems to be working! I've adjusted the darcy test based on your suggestion so that the porosity now linearly increases across the width of the model (everything else is the same). Once again, on the main branch this test crashes relatively quickly because the porosity blows up, but with the changes in this branch the test yields the expected behaviour (see screenshots below).

<img width="894" alt="image" src="https://github.com/user-attachments/assets/d01b68cc-0f29-4e7d-b663-f02ba83cee6f" />

<img width="902" alt="image" src="https://github.com/user-attachments/assets/5e94737e-f798-4095-ad2e-78512a89f14a" />

<img width="904" alt="image" src="https://github.com/user-attachments/assets/7d36baeb-2283-441a-a731-0f0b87733796" />

This is ready for another look through when you get a chance!

jdannberg commented on tests/darcy_outflow.prm:
```diff
@@ -0,0 +1,190 @@
+
+  subsection Function
+    set Variable names      = x,y,t
+    # set Function constants  = pi=3.1415926, x0=3750, y0=3750, c=1000
+    # set Function expression = 0.01 * exp(-((x-x0)*(x-x0)+(y-y0)*(y-y0))/(2*c*c)); 0.0; if(y>=7000, 1, 0)
```
These two lines can be deleted, right?

jdannberg commented on tests/darcy_outflow.prm:
```diff
@@ -0,0 +1,190 @@
+
+# Output the darcy velocity
+subsection Postprocess
+  set List of postprocessors          = velocity statistics, visualization
```
can you add the composition statistics postprocessor? Just because the value of the compositional fields is the thing we'd expect to change if things ever stopped working correctly.

jdannberg reviewed (commented):
The code looks good to me! I only have a few comments on the tests and wording.

jdannberg commented:
I think the only thing left is to update the description in the changelog, and then there's a problem with some of the tests:
Specifically, the `darcy_convection_step` says

> An error occurred in line <173> of file </__w/aspect/aspect/source/material_model/rheology/visco_plastic.cc> in function
>     aspect::MaterialModel::IsostrainViscosities aspect::MaterialModel::Rheology::ViscoPlastic<dim>::calculate_isostrain_viscosities(const aspect::MaterialModel::MaterialModelInputs<dim>&, unsigned int, const std::vector<double>&, const std::vector<double>&, const std::vector<unsigned int>&) const [with int dim = 2]
> The violated condition was: 
>     temperature_for_viscosity != 0
> Additional information: 
>     The temperature used in the calculation of the visco-plastic rheology
>     is zero. This is not allowed, because this value is used to divide
>     through. It is probably being caused by the temperature being zero
>     somewhere in the model. The relevant values for debugging are:
>     temperature (0), adiabatic_temperature_gradient_for_viscosity (0) and
>     pressure (0).

So it would be good to figure out what the problem is there. 

Other than that, the `no_dirichlet_on_outflow_melt` test fails, which makes sense because the PR changes how the composition boundary conditions are applied in cases with melt. But it would be good if you could check if the new test output looks reasonable. 

The `prescribed_velocity_boundary` only has a different number of calls to the boundary_velocity() function because we now do not call that any more for the compositions in compute_current_constraints () if no compositional fields are in the simulation.

And then of course you still have to update the test results for the new tests you made.

danieldouglas92 commented:
Tests that I have fixed/verified:

- [ ] Darcy Convection Step
- [ ] no_dirichlet_on_outflow_melt
- [ ] prescribed_velocity_boundary

gassmoeller commented:
@danieldouglas92: testers dont pass :-( let us know if you need help

danieldouglas92 commented:
@gassmoeller This should fix the test that was broken

jdannberg reviewed (approved):
Great! Thank you for fixing this!
